Allow task failure on disk imports

wfchandler opened the pull request:
Previously, when `oxide` initiated many concurrent uploads, network
congestion could cause packet loss and timeouts for some worker tasks.
While we added the undocumented `--parallelism` flag as a temporary
workaround, this required users to manually diagnose network issues
and adjust settings themselves.

This commit attempts to make disk imports more reliable by allowing the
job to continue when a subset of upload tasks have failed. When a worker
encounters a network error, it reports the failed chunk's file offset
back to the main task. The main task will retry these failed chunks
after completing the initial upload attempt. This creates natural
backpressure - as network congestion increases and more tasks fail, the
number of concurrent uploads automatically decreases to a sustainable
level.

This requires a switch to a single `mpmc` channel shared between all
worker tasks to avoid losing any buffered jobs when a task errors out.
We also move reading from the source file into the upload task for
improved parallelization.

wfchandler committed:
Allow task failure on disk imports

Previously, when `oxide` initiated many concurrent uploads, network
congestion could cause packet loss and timeouts for some worker tasks.
While we added the undocumented `--parallelism` flag as a temporary
workaround, this required users to manually diagnose network issues
and adjust settings themselves.

This commit attempts to make disk imports more reliable by allowing the
job to continue when a subset of upload tasks have failed. When a worker
encounters a network error, it reports the failed chunk's file offset
back to the main task. The main task will retry these failed chunks
after completing the initial upload attempt. This creates natural
backpressure - as network congestion increases and more tasks fail, the
number of concurrent uploads automatically decreases to a sustainable
level.

This requires a switch to a single `mpmc` channel shared between all
worker tasks to avoid losing any buffered jobs when a task errors out.
We also move reading from the source file into the upload task for
improved parallelization.

wfchandler commented:
I'm unable to add a unit test for the error handling without being able to replace mocks. I should resolve the merge conflicts on https://github.com/alexliesenfeld/httpmock/pull/108 and check on its status.

As a manual test, I made the following change to trigger all but one worker to fail:

```patch
diff --git a/sdk/src/extras/disk.rs b/sdk/src/extras/disk.rs
index 72d3d20..a8bf9dc 100644
--- a/sdk/src/extras/disk.rs
+++ b/sdk/src/extras/disk.rs
@@ -335,8 +335,9 @@ pub mod types {
             let (failed_tx, failed_rx) = flume::bounded(self.upload_thread_ct);
             let (resubmit_tx, resubmit_rx) = flume::bounded(self.upload_thread_ct);
 
-            for _ in 0..self.upload_thread_ct {
+            for idx in 0..self.upload_thread_ct {
                 let mut worker = UploadWorker {
+                    idx,
                     client: self.client.clone(),
                     disk: self.disk.clone(),
                     project: self.project.clone(),
@@ -354,6 +355,7 @@ pub mod types {
                 handles.push(tokio::spawn(async move {
                     while let Ok(offset) = rx.recv_async().await {
                         if let Err(err) = worker.upload_chunk(offset).await {
+                            eprintln!("worker {}: failed at offset {offset}", worker.idx);
                             let _ = failed_tx.send_async(offset).await;
                             return Err(err);
                         }
@@ -384,6 +386,7 @@ pub mod types {
 
             // Resubmit any failed offsets back to the remaining workers to retry.
             while let Ok(failed_offset) = failed_rx.recv_async().await {
+                eprintln!("main task: resubmitted offset {failed_offset}");
                 if resubmit_tx.send_async(failed_offset).await.is_err() {
                     // All worker tasks have failed.
                     break;
@@ -399,6 +402,7 @@ pub mod types {
                 }
             }
 
+            eprintln!("error count: {}", errors.len());
             // Only return an error if all worker tasks failed.
             if errors.len() == self.upload_thread_ct {
                 match errors.len() {
@@ -771,6 +775,7 @@ pub mod types {
     }
 
     struct UploadWorker {
+        idx: usize,
         client: Client,
         disk: Name,
         project: NameOrId,
@@ -783,6 +788,9 @@ pub mod types {
 
     impl UploadWorker {
         async fn upload_chunk(&mut self, offset: u64) -> Result<(), DiskImportError> {
+            if self.idx > 0 && offset % 12 == 0 {
+                return Err(DiskImportError::Other("synthetic error".into()));
+            }
             self.file.seek(SeekFrom::Start(offset))?;
             let n = match self
                 .file
```

This successfully imported a disk:

```
worker 3: failed at offset 0
worker 7: failed at offset 1572864
worker 5: failed at offset 3145728
worker 2: failed at offset 4718592
worker 6: failed at offset 6291456
worker 1: failed at offset 7864320
worker 4: failed at offset 11010048
main task: resubmitted offset 0
main task: resubmitted offset 1572864
main task: resubmitted offset 3145728
main task: resubmitted offset 4718592
main task: resubmitted offset 6291456
main task: resubmitted offset 7864320
main task: resubmitted offset 11010048
error count: 7
Done!
```

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -331,118 +331,92 @@ pub mod types {
+                    disk: self.disk.clone(),
+                    project: self.project.clone(),
+                    file_path: self.disk_info.file_path.clone(),
+                    file: File::open(&self.disk_info.file_path)?,
```
do we want to have the file open N times? It seems like doing concurrent `pread(2)` calls on the same fd would be fine. 

If it's not feasible to have a single fd, do we want to consider having a single task that does the file reading (I don't think we expect that local disk access is the thing that requires maximal parallelism).

wfchandler replied:
We could in theory share an `Arc<Mutex<File>>` between tasks, but that would block them from reading concurrently.

Moved file i/o into the main task to keep things simple.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -331,118 +331,92 @@ pub mod types {
+                    client: self.client.clone(),
+                    disk: self.disk.clone(),
+                    project: self.project.clone(),
+                    file_path: self.disk_info.file_path.clone(),
```
do we only need this for error messages?

wfchandler replied:
Yeah, but given that the user provided us with the file path I don't think there's much benefit to listing it in the error. Switched to a more generic error.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -331,118 +331,92 @@ pub mod types {
-                    Ok(())
-                }));
+                    // Signal the main task that the worker has sent all chunks successfully.
+                    drop(failed_tx);
```
perhaps this is picayune, but it's also important that there are no more chunks left to be sent.

wfchandler replied:
Updated the comment to be specific about this.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -331,118 +331,92 @@ pub mod types {
+                    while let Ok(offset) = rx.recv_async().await {
+                        if let Err(err) = worker.upload_chunk(offset).await {
+                            let _ = failed_tx.send_async(offset).await;
+                            return Err(err);
```
does the nature of the failure matter here or is it always the right thing to exit this task?

wfchandler replied:
Failure to send here would indicate that the main task has already exited, so we'd still want to bail out.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -331,118 +331,92 @@ pub mod types {
+                        for err in errors {
+                            msg += &format!("\n * {err}");
+                        }
+                        return Err(DiskImportError::Other(msg.into()));
```
this no longer seems right: the default use will produce 10 errors, mostly the same--is that correct?

wfchandler replied:
Updated to dedupe the error messages shown to the user.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -796,6 +770,64 @@ pub mod types {
+                .file
+                .by_ref()
+                .take(CHUNK_SIZE)
+                .read_to_end(&mut self.buf)
```
If I understand this correctly, this is blocking. I think this was less of a problem when we had a task doing the reading.

wfchandler replied:
Moved to async file i/o, good catch!

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -796,6 +770,64 @@ pub mod types {
+
+            if n == 0 {
+                return Ok(());
+            }
```
how would we encounter this case?

wfchandler replied:
We've reached the end of the file, but this is now in the main task.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -796,6 +770,64 @@ pub mod types {
+                self.client
+                    .disk_bulk_write_import()
+                    .disk(self.disk.clone())
+                    .project(self.project.clone())
```
I don't believe the `clone()` calls are necessary, but it's fine either way.

wfchandler replied:
Thanks, fixed. It's surprising that I need to deref a `Name` to avoid a clone, not sure how to make that more ergonomic, though.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -796,6 +770,64 @@ pub mod types {
+
+            self.upload_progress.fetch_add(n as u64, Ordering::Relaxed);
+            self.progress_tx
+                .send_replace(self.upload_progress.load(Ordering::Relaxed));
```
```suggestion
            let x = self.upload_progress.fetch_add(n as u64, Ordering::Relaxed) + n as u64;
            self.progress_tx
                .send_replace(x);
```

does that work?

wfchandler replied:
Thinking about this a little more we don't need the atomic at all, and can just use the `send_modify` method on the channel directly.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -796,6 +770,64 @@ pub mod types {
+            self.upload_progress.fetch_add(n as u64, Ordering::Relaxed);
+            self.progress_tx
+                .send_replace(self.upload_progress.load(Ordering::Relaxed));
+            self.buf.clear();
```
Is this out of an abundance of caution?

wfchandler replied:
`read_to_end` will append to the end of the `Vec`.  We need to clear it so we can re-use the buffer.

ahl replied:
got it! thanks

ahl commented:
> Previously, when `oxide` initiated many concurrent uploads, network congestion could cause packet loss and timeouts for some worker tasks.

Do we know that was the proximate cause?

> While we added the undocumented `--parallelism` flag as a temporary workaround, this required users to manually diagnose network issues and adjust settings themselves.

While that's true, we also just made up some arbitrary number for the permissible parallelism.

ahl commented:
Before proceeding (or in this PR), can we please switch to a more reasonable tokio configuration? In particular, we're creating a thread  per core and that doesn't seem reasonable for this CLI.

Does disk upload even require multiple threads or would a single-threaded executor suffice?

wfchandler commented:
> Before proceeding (or in this PR), can we please switch to a more reasonable tokio configuration? In particular, we're creating a thread per core and that doesn't seem reasonable for this CLI.
> 
> Does disk upload even require multiple threads or would a single-threaded executor suffice?

Testing out disk uploads from a 16-core VM into colo, the fastest runs for with `#[tokio::main(flavor = "current_thread")]` are on par with the multi-threaded executor, but the slowest requests were a fair bit slower.

I ran five uploads of a 3 GiB Debian image against this PR with the different executors, as well as main:

| this PR - current_thread | this PR - multi_thread | main |
| --- | --- | --- | 
| 29.941s | 28.066s | 30.672s | 
| 32.554s | 28.770s | 31.322s |
| 35.911s | 31.481s | 32.162s |
| 39.440s | 35.060s | 33.443s |
| 40.344s | 36.796s | 37.670s |
| AVG 35.638s | AVG 32.035s | AVG 33.054s |


Unsurprisingly `oxide version` shows a signficant improvement in performance:

```
$ hyperfine -L oxide './oxide-st, ./oxide-main' -w3 '{oxide} version'
Benchmark 1: ./oxide-st version
  Time (mean ± σ):      14.5 ms ±   0.6 ms    [User: 9.2 ms, System: 5.3 ms]
  Range (min … max):    13.3 ms …  17.9 ms    187 runs
 
Benchmark 2:  ./oxide-main version
  Time (mean ± σ):      22.1 ms ±   1.0 ms    [User: 11.6 ms, System: 13.3 ms]
  Range (min … max):    20.2 ms …  26.4 ms    127 runs
 
Summary
  ./oxide-st version ran
    1.52 ± 0.10 times faster than  ./oxide-main version
```

A simple command that does reach out to the network shows a smaller benefit:

```
$ hyperfine -L oxide './oxide-st, ./oxide-main' '{oxide} instance list --project=will'
Benchmark 1: ./oxide-st instance list --project=will
  Time (mean ± σ):     156.4 ms ±   5.8 ms    [User: 12.8 ms, System: 7.8 ms]
  Range (min … max):   149.4 ms … 174.7 ms    18 runs
 
Benchmark 2:  ./oxide-main instance list --project=will
  Time (mean ± σ):     163.6 ms ±   8.6 ms    [User: 14.9 ms, System: 17.5 ms]
  Range (min … max):   153.5 ms … 191.6 ms    18 runs
 
Summary
  ./oxide-st instance list --project=will ran
    1.05 ± 0.07 times faster than  ./oxide-main instance list --project=will
```

However, this mostly washes out when run from my more distant home network, though CPU usage remains lower:

```
$ hyperfine -M5 -L oxide './oxide-st,./oxide-main' '{oxide} instance list --project=will'
Benchmark 1: ./oxide-st instance list --project=will
  Time (mean ± σ):     541.2 ms ±  15.3 ms    [User: 13.9 ms, System: 13.8 ms]
  Range (min … max):   515.1 ms … 554.2 ms    5 runs
 
Benchmark 2: ./oxide-main instance list --project=will
  Time (mean ± σ):     544.5 ms ±   3.4 ms    [User: 16.4 ms, System: 19.7 ms]
  Range (min … max):   541.0 ms … 549.6 ms    5 runs
 
Summary
  ./oxide-st instance list --project=will ran
    1.01 ± 0.03 times faster than ./oxide-main instance list --project=will
```

Overall not a big benefit, but it's measurable. I'll open a separate PR to use the `current_thread` executor unless we're performing a disk import.

wfchandler committed:
Replace synchronous file i/o

Use the tokio for reading from the file to minimize blocking operations.

wfchandler committed:
Remove redundant atomic progress variable

The `watch` channel gives us a thread-safe way to communicate changes
between tasks. Just use the `send_modify` method to increment the value
and get rid of the unneeded `AtomicU64`.

wfchandler committed:
Read file in main task

To avoid opening the file <WORKER_TASK_CT> times, read the raw data from
the main task and ship that to the workers.

wfchandler committed:
Dedupe error messages

Worker tasks are likely to encounter the same error when uploading
chunks. Rather than listing the same error N times, dedupe them and show
only unique errors.

wfchandler committed:
Make upload_thread_ct NonZeroUsize

It is invalid to set upload_thread_ct to zero. Update the API to enforce
this constraint.

wfchandler committed:
Rename upload_thread_ct to upload_task_ct

The `upload_thread_ct` variable name is inaccurate, as we're creating
tokio tasks to perform uploads, not OS threads.

Correct the name to `upload_task_ct`.

wfchandler committed:
Note that no chunks remain in comment

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -328,56 +325,55 @@ pub mod types {
-                        progress_tx.send_replace(upload_progress.load(Ordering::Relaxed));
+                    while let Ok(chunk) = rx.recv_async().await {
+                        if let Err(e) = worker.upload_chunk(&chunk).await {
+                            let _ = failed_tx.send_async(chunk).await;
```
if this fails... it's kind of a huge problem, right? like: we've just silently corrupted the disk image on the rack

wfchandler replied:
No, this is equivalent to [previous while let Some](https://github.com/oxidecomputer/oxide.rs/blob/e633564d2fb16a6a8863f458f4530d0b7f4eba56/sdk/src/extras/disk.rs#L346) loop we used with the tokio channel.  `flume` treats the sender droping [as an error](https://docs.rs/flume/latest/flume/enum.RecvError.html). This can happen in one of two situations:

1. The main task has exited early, in which case we won't finalize the disk.
2. The main task has sent all chunks to the channel and [dropped the Sender](https://github.com/oxidecomputer/oxide.rs/blob/49a7c63aab56c869584a204faecfdad917a29eff/sdk/src/extras/disk.rs#L410), and there's no more chunks to upload.

ahl replied:
Awesome; thanks!

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -328,56 +325,55 @@ pub mod types {
+                    drop(failed_tx);
+
+                    while let Ok(chunk) = resubmit_rx.recv_async().await {
+                        worker.upload_chunk(&chunk).await?;
```
why are we not putting failures back on the failed_tx?

wfchandler replied:
This is something I went back and forth on.

Workers close `failed_tx` to indicate they has finished all requests in `tx`. If we keep `failed_rx` open for resubmits then we deadlock.

We could add yet another channel for failed resubmits, but given that we would have at most <WORKER_CT - 1> retries, I didn't think it was worth the extra complexity.

ahl replied:
let's not deadlock! this is already better than it was so let's just proceed

wfchandler commented:
Quick sanity check that upload performance is inline with where we were prior to these changes and the runtime switch, 7a92c41a. On average slightly faster, with one slow outlier.

```
$ hyperfine -L oxide './oxide-7a92c41a,./oxide-latest' 'bash -c "{oxide} disk import --project will --description BenchRTs --path ~/alpine-3.21.0-nocloud-amd64.raw --disk bench-$(date +%s%N) --disk-block-size 512"'
Benchmark 1: bash -c "./oxide-7a92c41a disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512"
  Time (mean ± σ):      7.103 s ±  0.846 s    [User: 0.619 s, System: 1.131 s]
  Range (min … max):    5.507 s …  8.304 s    10 runs
 
Benchmark 2: bash -c "./oxide-latest disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512"
  Time (mean ± σ):      6.811 s ±  1.572 s    [User: 0.641 s, System: 1.090 s]
  Range (min … max):    5.130 s … 10.090 s    10 runs
 
Summary
  bash -c "./oxide-latest disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512" ran
    1.04 ± 0.27 times faster than bash -c "./oxide-7a92c41a disk import --project will --description BenchRTs --path ~/alpine.raw --disk bench-$(date +%s%N) --disk-block-size 512"
```

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -389,60 +385,68 @@ pub mod types {
                 }
 
                 // If the chunk we just read is all zeroes, don't POST it.
```
```suggestion
                // If the chunk we just read is all zeroes, we don't need to upload it.
```

wfchandler replied:
👍

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -389,60 +385,68 @@ pub mod types {
 
-                offset += CHUNK_SIZE;
-                i += 1;
+                offset += n as u64;
```
this is interesting; we do we expect partial reads other than at the end of the disk image?

wfchandler replied:
As you say, in practice we should only have a short read at the end of the file. I thought it would be better to keep `offset` in sync with our position in the file just in case a future change somehow invalidated this assumption.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -389,60 +385,68 @@ pub mod types {
-                    self.upload_progress.fetch_add(n as u64, Ordering::Relaxed);
-                    self.progress_tx
-                        .send_replace(self.upload_progress.load(Ordering::Relaxed));
+                    // Bump the progress bar here to make it consistent.
```
It took me a minute to remind myself what this meant; for your consideration--feel free to ignore
```suggestion
                    // Bump the progress bar here. We don't need to send this chunk so we've completed the relevant work.
```

wfchandler replied:
Good improvement, thanks.

ahl reviewed (commented):
looks good

wfchandler committed:
Apply suggestions from code review

Co-authored-by: Adam Leventhal <ahl@oxide.computer>

wfchandler commented:
> > Previously, when `oxide` initiated many concurrent uploads, network congestion could cause packet loss and timeouts for some worker tasks.
> 
> Do we know that was the proximate cause?

We know it was the for the [failures we saw in CI](https://github.com/oxidecomputer/omicron/issues/6771#issuecomment-2430681859) last fall. Still working on root cause for https://github.com/oxidecomputer/oxide.rs/issues/993, but given that reducing the number of upload tasks significantly reduces their rate of failure, it seems likely that this PR will help with that issue as well.

ahl reviewed (approved):
looks good! maybe update copyrights as you see fit?

wfchandler committed:
Update copyright year
